### PR TITLE
Implement BPF-based masquerading for IPv6 (rebased)

### LINKED ISSUE #14350: Implement BPF based masquerading for IPv6 traffic
Follow up from #14124

### PR DESCRIPTION
**Pull request by qmonnet**
Summary of commits:
1. Preparatory: Refactor `ct_is_reply4` as a macro to facilitate implem. of IPv6 counterpart.
2. Preparatory (verifier): Change `ipv6_addrcmp()` into `ipv6_addr_equals()`.
3. Preparatory (verifier): Make `DEFINE_U*()` macros use `.data` section for declarations.
4. Preparatory: Rename `ENABLE_MASQUERADE` as `ENABLE_MASQUERADE_IPV4`.
5. Implement the datapath changes by following the IPv4 implementation.
6. Add ICMP support to the new `ct_extract_ports6()`, refactor to get closer to IPv4's version.
7. Add `ENABLE_MASQUERADE_IPV6` to options lists for complexity tests.
8. Preparatory: Refactor `InitBPFMasqueradeAddrs` to facilitate implem. of IPv6 counterpart.
9. Implement agent changes for the flags and macros.
10. Forbid IPv6 BPF masquerading with host firewall, until we look deeper into the datapath aspects.
11. Update documentation.
12. 13\. 14\. Extend/adjust the tests to cover IPv6 BPF masquerading.

See commits for details.

Fixes: https://github.com/cilium/cilium/issues/14350.

```release-note
Implement BPF-based masquerading for IPv6
```

This is a rebase of https://github.com/cilium/cilium/pull/18694 from Paul.

### CODE REVIEW

**Comment by qmonnet** [comment]
test-1.16-4.19

**Comment by qmonnet** [comment]
/test-1.24-5.4

**Comment by qmonnet** [comment]
/test-1.16-4.19

**Comment by qmonnet** [comment]
/test-1.26-net-next

**Comment by qmonnet** [comment]
/test-1.26-net-next

**Comment by qmonnet** [comment]
/test-1.26-net-next

**Comment by qmonnet** [comment]
/test-1.26-net-next

File: bpf/lib/nat.h
Comment (julianwiedmann) [inline bpf/lib/nat.h]:
```suggestion
			l4_off = ETH_HLEN + ipv6_hdrlen(ctx, &tuple.nexthdr);
```


to extract the L4 protocol into the CT tuple (for the map lookup), and to avoid that `ipv6_hdrlen()` overwrites the nexthdr field in the IPv6 header.

Comment (qmonnet) [reply bpf/lib/nat.h]: Right, thanks! Updated

File: bpf/lib/nat.h
Comment (julianwiedmann) [inline bpf/lib/nat.h:1684]: Could you also remove the `nodeport_uses_dsr()` forward declaration at the top of the file? :pray: 

(now I'm looking forward to this PR even more haha)

Comment (qmonnet) [reply bpf/lib/nat.h:1684]: Yep, done! Thanks

**Comment by qmonnet** [comment]
I added `dont-merge/preview-only`, not because it's a preview really, but just so that we don't merge by mistake until the last commit (to test `conformance-e2e`) has been removed.

File: Documentation/network/concepts/masquerading.rst
Comment (zacharysarah) [inline Documentation/network/concepts/masquerading.rst]: I'd like to see "currently" removed, but I'm willing to defer it as out of scope for the PR.

**Review by zacharysarah — APPROVED** [review]
LGTM with one fond wish 🥺

File: Documentation/network/concepts/masquerading.rst
Comment (PKizzle) [reply Documentation/network/concepts/masquerading.rst]: So are we not going to see IPv6 support for ip-masq-agent? Or why do you want to remove "currently"?

Comment (qmonnet) [reply Documentation/network/concepts/masquerading.rst]: My understanding from Sarah's feedback is that having “currently” in the docs is not a good practice - you never really when “currently” refers to (it refers to the date of writing obviously, but when reading the docs, how do you figure this one out?).

Sarah, I'll remove that word for the next iteration. Thanks for the review!

As for the ip-masq-agent with IPv6, it is coming [as a follow-up PR](https://github.com/cilium/cilium/pull/23219), it's _currently_ waiting for this one to be merged.

**Review by tklauser — APPROVED** [review]
CI bits LGTM sans the test `DO NOT MERGE` commit which I assume will be dropped before merging. I've also briefly skimmed through the other bits in this PR. Nice work!

File: bpf/bpf_host.c
Comment (julianwiedmann) [inline bpf/bpf_host.c:1157]: The asymmetry with the IPv6 path concerns me. But that's not something introduced by your PR :).

File: bpf/lib/conntrack.h
Comment (julianwiedmann) [inline bpf/lib/conntrack.h:402]: This needs to handle SCTP now as well, correct?

**Review by mhofstetter — APPROVED** [review]
sig-agent part looks good!

File: bpf/lib/conntrack.h
Comment (julianwiedmann) [inline bpf/lib/conntrack.h]: What does it mean in practice if we don't have ICMPv6 support here? Looks like we wouldn't be able to classify traffic as belonging to an inbound connection, and miss to skip SNAT?

File: bpf/lib/nat.h
Comment (julianwiedmann) [inline bpf/lib/nat.h:1668]: Did you check whether this is still in sync with `snat_v4_prepare_state()` ? Right now it's a bit difficult to check whether we have all the corresponding cases covered.

I expected them to look slightly different because of EgressGW, but not sure that explains the whole diff :).

File: bpf/lib/nat.h
Comment (julianwiedmann) [inline bpf/lib/nat.h]: Note that `ipv6_hdrlen()` can return errors.

File: bpf/lib/conntrack.h
Comment (julianwiedmann) [inline bpf/lib/conntrack.h]:
```suggestion
		if (l4_load_ports(ctx, off, &tuple->dport) < 0)
			return DROP_CT_INVALID_HDR;
```

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
lgtm overall, thanks for working on this Quentin!

One error-handling case that needs to be resolved, and a few clarifying questions below :).

File: Documentation/network/concepts/masquerading.rst
Comment (qmonnet) [reply Documentation/network/concepts/masquerading.rst]: Now updated in the docs.

File: bpf/bpf_host.c
Comment (qmonnet) [reply bpf/bpf_host.c:1157]: What do you mean, having two different paths `ENABLE_MASQUERADE_IPV4` and `...IPV6`? I can't see how to avoid it with the options supported by the Agent, given that you can turn on/off masquerading for one IP version or the other independently :thinking:.

File: bpf/lib/conntrack.h
Comment (qmonnet) [reply bpf/lib/conntrack.h:402]: Sounds correct, I added it.

File: bpf/lib/conntrack.h
Comment (qmonnet) [reply bpf/lib/conntrack.h]: Hmm good question, yes possibly. I added commit https://github.com/cilium/cilium/pull/23165/commits/410aac7ec149a711ae43b954fa837328c79e6853 to refactor a bit based on what Maxim did for IPv4, and to add ICMP support here as well. Can you please take a look?

File: bpf/lib/conntrack.h
Comment (qmonnet) [reply bpf/lib/conntrack.h]: Updated, thanks.

File: bpf/lib/nat.h
Comment (qmonnet) [reply bpf/lib/nat.h:1668]: I did check. I came up with [a commit](https://github.com/cilium/cilium/commit/243534ea087b84e08c09df3ab10380d31f76df9e) to update `snat_v6_needed()` with respects to `snat_v4_prepare_state()`, but so far I've not included it in the PR.

I noted that the function evolved mostly with the following commits:

  - cdfc30d ("bpf: egressgw: sync logic to determine if destination is outside cluster")
    Moving the `ct_is_reply()` check above the `IPV4_SNAT_EXCLUSION_DST_CIDR` logics makes sense to me, so we can probably apply the non-EGW part of this commit to `snat_v6_needed()`.

  - https://github.com/cilium/cilium/commit/453299601c8c5747deb0b004325db796a2fd31f9 ("bpf: nat: always track egress gateway connections")
    This one is some refactoring which is not necessary at the moment for IPv6, but it seems trivial to apply the few bits touching the relevant function to align with IPv4.

  - https://github.com/cilium/cilium/commit/59ac84f65d2c14fdf8038b29a53b2d1a87f18598 ("bpf: nat: consolidate snat_v4_needed() state into ipv4_nat_target")
    Same.

I probably missed some older commits in the list, but https://github.com/cilium/cilium/commit/243534ea087b84e08c09df3ab10380d31f76df9e brings the two versions (IPv4 and IPv6) pretty close to one another. I'm happy to rework it and add it to the PR.

Let's maybe sync sometimes on Slack/Zoom?

File: bpf/lib/nat.h
Comment (qmonnet) [reply bpf/lib/nat.h]: Good catch. Not addressed yet, I'll need to change the return type of the function to be able to pass an error. Let's see what we do of `snat_v6_needed()` first.

File: bpf/bpf_host.c
Comment (julianwiedmann) [reply bpf/bpf_host.c:1157]: It looks like we have special handling for the "no IPv4 BPF masq" case in HostFW, but not yet in the IPv6 path.

I haven't tracked it down further at all, but would speculate that this is because the corresponding IPv6 HostFW path right now **only** has the "no BPF masq" handling. And would need to be extended in similar fashion as IPv4 to also handle BPF masq.

Needs further digging whether that handling for BPF masq in HostFW is an optional optimization, or strictly required.

File: bpf/lib/conntrack.h
Comment (julianwiedmann) [reply bpf/lib/conntrack.h]: thanks, will stare at it! @gentoo-root FYI

File: bpf/lib/nat.h
Comment (julianwiedmann) [reply bpf/lib/nat.h]: FWIW, I think it's fair to only bubble up the error as far as we do in the IPv4 path. And then improve things for both paths in a follow-on.

File: bpf/lib/nat.h
Comment (julianwiedmann) [reply bpf/lib/nat.h:1668]: Ah nice, thank you! I would need to stare a bit more at your current code to understand if something is genuinely missing, not sure how quickly we could fit in that session :/.

So if you already lined up a sync-up commit for a follow-on, I would probably be fine with merging as-is. And then let's review separately if the sync-up fixes any issues (and needs to go into v1.14), or can wait until v1.15.

**Comment by qmonnet** [comment]
/test-1.26-net-next

File: bpf/lib/nat.h
Comment (qmonnet) [reply bpf/lib/nat.h]: OK, I'll address by skipping the `ct_is_reply6()` if we get a wrong offset, and leave a comment to improve that in the future.

File: bpf/lib/nat.h
Comment (qmonnet) [reply bpf/lib/nat.h:1668]: Alright, let's keep the current version and make it on par with IPv4 as a follow-up. I'll file an issue to list the different follow-ups to do for this PR.

File: bpf/bpf_host.c
Comment (qmonnet) [reply bpf/bpf_host.c:1157]: OK, right, let's disable BPF IPv6 masquerading in conjunction with Host Firewall until we figure out the details here. Thanks a lot for the review and feedback!

File: bpf/lib/nat.h
Comment (qmonnet) [reply bpf/lib/nat.h:1668]: https://github.com/cilium/cilium/issues/26074

**Review by viktor-kurchenko — APPROVED** [review]

File: bpf/lib/conntrack.h
Comment (julianwiedmann) [reply bpf/lib/conntrack.h]: looks good, ty!

File: bpf/bpf_host.c
Comment (julianwiedmann) [reply bpf/bpf_host.c:1157]: Thanks! For the record, this is tracked in https://github.com/cilium/cilium/issues/26074.

**Review by julianwiedmann — APPROVED** [review]
Checks out, and the remaining follow-on items are ok. Thank you!

**Comment by joestringer** [comment]
> the remaining follow-on items are ok

What are the remaining follow-on items? Is there an issue to track them?

**Review by nathanjsweet — APPROVED** [review]

**Review by rgo3 — APPROVED** [review]

**Comment by qmonnet** [comment]
/test-1.26-net-next

**Comment by qmonnet** [comment]
Conformance E2E with changes from this PR to be checked at https://github.com/cilium/cilium/pull/26046 - Currently red because the connectivity tests involving 1.1.1.1 are failing (likely for an external reason, as I've seen them fail on multiple PRs/branches)

**Comment by qmonnet** [comment]
CI was all green at last, but https://github.com/cilium/cilium/commit/e4d277cfe380d2ac9f0ef01451a557e13ccf90c7 created a conflict on `bpf/tests/tc_egressgw_snat.c` :upside_down_face:.

**Comment by qmonnet** [comment]
CI's all green, apart from checkpatch's false positives (I checked them). Updated Conformance E2E [on the test PR](https://github.com/cilium/cilium/pull/26046) is all green, too. All reviews are in. We're good to go! :tada:

**Pull request merged**
